docs/ralph/danube-features.md

```diff
@@ -0,0 +1,254 @@
+    }
+}
+
+TxScript ChainedSwapToken(tokenPair12: Swap, tokenPair23: Swap, token1: ByteVec, token2: ByteVec) {
+    let caller = callerAddress!()
+    tokenPair12.swap{caller -> token1: 5}(caller, token1, 5)
+    tokenPair23.swap{caller -> token2: 5}(caller, token2, 5)
+}
+```
+
+In the example above, the TxScript `ChainedSwapToken` performs two swap operations in sequence. First, it swaps `token1` for `token2` using the `tokenPair12` swap contract, and then it uses the received `token2` to swap for `token3` using the `tokenPair23` swap contract. This chaining of asset outputs would not be possible before the Danube upgrade, as the tokens received from the first swap would not be available for use in the second swap within the same transaction.
+
+### Caller Address
+
+The ability to determine the caller of a function is important for implementing proper access control and security measures in smart contracts. Before the Danube upgrade, there are two built-in functions to determine the function caller: `callerAddress!()` and `callerContractId!()`.
+
+`callerAddress!()` works in the following ways:
+
+1. Return the transaction caller when called in a TxScript or a contract function directly called by TxScript
+2. Return the address of the calling contract when called in a contract function that is not called by the TxScript
+
+`callerContractId!()` fails in scenario 1 (no calling contract) but works exactly like `callerAddress!()` in scenario 2, except that it returns the contract ID instead of the address of the calling contract.
+
+Despite being very useful, there are two limitations with how these two built-in functions work:
+
+First, when a function calls another function within the same contract, `callerAddress!()` in the called function returns the contract's own address. This limits dApps that need to identify the external caller of the current contract.
+
+Second, in a chain of function calls, `callerAddress!()` returns the immediate caller. However, sometimes functions want to preserve the identity of the caller and pass it along to the next function in the call chain. This is particularly useful for contracts implementing the routing patterns where maintaining the identity of the original caller is essential for proper access control and authorization.
+
+The Danube upgrade provides solutions to address both of these limitations.
+
+#### External Caller Address
+
+
+The Danube upgrade introduces the `externalCallerAddress!()` and `externalCallerContractId!()` built-in functions. These two functions return the address and contract ID of the first external caller from outside the current contract, regardless of how many internal function calls have occurred.
+
+The following example demonstrates how `externalCallerAddress!()` works, `externalCallerContractId!()` works in a similar way:
+
+```rust
+Contract ExternalContract(internal: InternalContract) {
+    pub fn callInternal() -> (Address, Address) {
+        return internal.call()
```

Comment:
[Nit]: Perhaps we can use the `assert!` statement like this:
```
let (caller, externalCaller) = internal.call()
assert!(caller == contractAddress!(internal), 0)
assert!(externalCaller == selfAddress!(), 0)
```

Reply:
Great idea, thanks.